### PR TITLE
Solution to coffee shops challenge.

### PR DESCRIPTION
**Pull request by vladstanescu94**
## How to run:

1. Install Python 3.x
2. Make python virtual environment in project directory Ex:
   - Mac OS and Linux: python3 -m venv env
   - Windows: py -m venv env
3. Once you’ve created a virtual environment, you may activate it:
   - On Windows run: env\Scripts\activate.bat
   - On Mac OS / Linux run: source env/bin/activate
4. Install dependencies with: pip install -r requirements.txt
5. Run tests with: python -m unittest
6. Run coffee_shop_finder.py ex:
   - python coffee_shop_finder.py <user_latitude> <user_longitude> <csv_data_URL>
   - python coffee_shop_finder.py 47.6 -122.4 https://raw.githubusercontent.com/vladstanescu94/agilefreaks-challenge/develop/Resources/coffee_shops.csv

### CODE REVIEW

File: Models/coordinates.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    def __init__(self, latitude, longitude):
+        try:
+            self.latitude = float(latitude)
+            self.longitude = float(longitude)
+        except (ValueError, TypeError) as error:
+            print(f"Invalid Coordinates Provided: {error}")
+            sys.exit(1)
+
+    def compute_distance_to(self, target_coordinates):
+        return round(sqrt((self.latitude - target_coordinates.latitude)**2 + (self.longitude - target_coordinates.longitude)**2), 4)
```
Comment (RalucaMihaela) [inline Models/coordinates.py:15]: This line is too long and a bit hard to follow. Can you split each functionality into smaller pieces?

File: coffee_shop_finder.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from Services.user_service import UserService
+from Services.csv_service import CSVService
+from Services.coffee_shop_service import CoffeeShopService
+from Utils.custom_csv_error import CSVServiceError
+
+
+def main():
+    user_args = sys.argv[1:]
+    user_service = UserService(args=user_args)
+    csv_service = CSVService()
```
Comment (alexandru-calinoiu) [inline coffee_shop_finder.py:11]: Good idea on extracting the logic to handle input in services.

File: coffee_shop_finder.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    try:
+        user_data = user_service.get_user_data()
+        csv_data = csv_service.get_csv_data(user_data.csv_url)
+    except (IndexError, CSVServiceError) as error:
```
Comment (alexandru-calinoiu) [inline coffee_shop_finder.py:16]: Is great that you handled specific errors instead of generic ones,

Why did you choose to go with IndexError instead of a custom error class like the CSVServiceError?

File: Services/csv_service.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+class CSVService:
+
+    def get_csv_data(self, url):
+        response = self._get_url_response(url)
+        csv_data = self._get_csv_content(response)
+        return csv_data
+
+    def _get_url_response(self, url):
+        response = requests.get(url, stream=True)
+        if response.status_code == 200:
```
Comment (alexandru-calinoiu) [inline Services/csv_service.py:17]: 200 is a well know error code but still feels like a [magic number](https://en.wikipedia.org/wiki/Magic_number_%28programming%29)

Can you think of an alternative here that can make things even more clear?

File: Services/coffee_shop_service.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    def generate_coffee_shops(self, csv_data):
+        coffee_shops = []
+        for row in csv_data:
+            try:
+                shop_name = row[0]
+                shop_coordinates = Coordinates(row[1], row[2])
+                shop = CoffeeShop(shop_name, shop_coordinates)
+                coffee_shops.append(shop)
+            except:
+                print(f"Invalid data ... Skipping over row {row} ")
```
Comment (alexandru-calinoiu) [inline Services/coffee_shop_service.py:16]: Is great that you try to parse as many results as possible,

The problem is that you are mixing up presentation with business logic. If you service is to be used from a web app for example the print here won't be very useful to the end user.

File: Tests/test_csv_service.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    def setUp(self):
+        self.csv_service = CSVService()
+
+    def tearDown(self):
+        del self.csv_service
+
+    def test_get_csv_data_raises_exception_when_invalid_url_provided(self):
+        url = "https://raw.githubusercontent.com/vladstanescu94/agilefreaks-challenge/develop/Resources/"
+
+        with self.assertRaises(CSVServiceHTTPError):
```
Comment (alexandru-calinoiu) [inline Tests/test_csv_service.py:17]: Using an external service in your unit tests will make them brittle and slow.

File: Tests/test_user_service.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import unittest
+
+from Models.user_data import UserData
+from Models.coordinates import Coordinates
+from Services.user_service import UserService
+
+
+class UserServiceUnitTest(unittest.TestCase):
```
Comment (alexandru-calinoiu) [inline Tests/test_user_service.py:8]: :+1:  for having unit tests that cover as much of the code paths as possible.

File: Tests/test_coordinates.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import unittest
+from Models.coordinates import Coordinates
+
+
+class CoordinatesUnitTests(unittest.TestCase):
+    def test_compute_distance_to_returns_expected_value(self):
```
Comment (alexandru-calinoiu) [inline Tests/test_coordinates.py:6]: :+1: for testing the things that matter, I hate it when I see tests asserting setters and getters

**Review by alexandru-calinoiu — COMMENTED** [review]
Very good solution, letf a couple of small comments, let me know what you think.

File: Services/csv_service.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        content = []
+        wrapper = csv.reader(codecs.iterdecode(response.iter_lines(), 'utf-8'), delimiter=',')
+        for record in wrapper:
+            content.append(record)
+
+        self._validate_data(content)
+        return content
+
+    def _validate_data(self, data):
+        if len(data) < 1:
```
Comment (RalucaMihaela) [inline Services/csv_service.py:32]: Is there a better way to check if your content is empty? Instead of using the magic number 1?

**Review by RalucaMihaela — COMMENTED** [review]
Great solution to the problem, I left also some small comments.

File: coffee_shop_finder.py (hunk 3 above)
Comment (vladstanescu94) [reply coffee_shop_finder.py:16]: I'm expecting index error when parsing user input data, so it's not a CSVServiceError but it could be a UserServiceError.

File: Services/csv_service.py (hunk 4 above)
Comment (vladstanescu94) [reply Services/csv_service.py:17]: We can use an enum of possible status codes or a module only with constants ex HTTP_SUCCESS = 200

File: Services/csv_service.py (hunk 9 above)
Comment (vladstanescu94) [reply Services/csv_service.py:32]: Yes, the correct solution i think is -> if not data: do stuff

File: Models/coordinates.py (hunk 1 above)
Comment (vladstanescu94) [reply Models/coordinates.py:15]: I can split the latitude part and the longitude part of the formula in separate variables. The rounding of the result can be a separate line as well.

File: Models/coordinates.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+from math import sqrt
+
+
+class Coordinates:
+    def __init__(self, latitude, longitude):
+        try:
+            self.latitude = float(latitude)
+            self.longitude = float(longitude)
+        except (ValueError, TypeError) as error:
+            print(f"Invalid Coordinates Provided: {error}")
```
Comment (AlexandraDamaschin) [inline Models/coordinates.py:11]: Is it ok to have hardcoded values in classes or there is something else we can do with them?

File: Services/user_service.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+    def get_user_data(self):
+        self.validate_user_args()
+        user_coordinates = Coordinates(self.args[0], self.args[1])
+        user_url = self.args[2]
+        return UserData(user_coordinates, user_url)
+
+    def validate_user_args(self):
+        if len(self.args) < 3:
+            raise IndexError("Error - must provide 3 arguments as input: latitude logitude csv_url")
```
Comment (AlexandraDamaschin) [inline Services/user_service.py:17]: Really good that you are catching different errors and let the user know if something went wrong, but how could we do to have all the errors in one place and just use them from there?

File: how_to_run.md [hunk 12]
```diff
@@ -0,0 +1,16 @@
+# Coffee Shop Finder
+
+## How to run:
+
+1. Install Python 3.x
+2. Make python virtual environment in project directory Ex:
```
Comment (AlexandraDamaschin) [inline how_to_run.md:6]: Really cool that you added this how to run for all OS :thumbsup:

File: Services/user_service.py (hunk 11 above)
Comment (vladstanescu94) [reply Services/user_service.py:17]: I guess we can have a module to handle just validation

File: Models/coordinates.py (hunk 10 above)
Comment (vladstanescu94) [reply Models/coordinates.py:11]: Hardcoded values are bad, but i don't really understand about what values do you talk here? If it's about my force cast to float i can see why it's not ok to have it here.  If you are talking about the print message i guess i could make a custom exception that has a custom message.

Comment (AlexandraDamaschin) [reply Models/coordinates.py:11]: Nice! I was referring to both and I really like your idea about having a custom exception with a message :100:

File: Services/csv_service.py (hunk 4 above)
Comment (alexandru-calinoiu) [reply Services/csv_service.py:17]: Indeed, further more one such module is already defined in python https://docs.python.org/3/library/http.html
